test/xbgpu/test_engine.py

```diff
@@ -1452,3 +1465,99 @@ def get_stream_status(stream_name: str) -> bool:
+        Also ensure that data is completely received for streams that did not
+        receive a `?capture-stop` request.
+        """
+        n_batches = heap_accumulation_threshold[0]
+        # Naive list concatention using '+' not used below as mypy was not happy
+        # lists of two different types could be added together.
+        all_stream_names = list(map(lambda output: output.name, chain(corrprod_outputs, beam_outputs)))
+        rng = np.random.default_rng(seed=1)
+        # Select random subset of output data products to issue `?capture-stop` to
+        n_streams_to_stop = rng.integers(1, len(all_stream_names))
+        stop_ids = rng.choice(len(all_stream_names), n_streams_to_stop, replace=False)
+        stopped_stream_names = [all_stream_names[stop_id] for stop_id in stop_ids]
+
+        capture_stop_requests = [("capture-stop", stopped_stream) for stopped_stream in stopped_stream_names]
+
+        capture_stop_chunk_id = 10  # Arbitrarily chosen
+        _ = self._patch_get_in_item(monkeypatch, capture_stop_chunk_id, client, requests=capture_stop_requests)
+        timestamp_step = n_samples_between_spectra * n_spectra_per_heap
+
+        def heap_factory(batch_index: int, present: np.ndarray) -> list[spead2.send.HeapReference]:
+            timestamp = batch_index * timestamp_step
+            data = np.full((n_channels_per_substream, n_spectra_per_heap, N_POLS, COMPLEX), 10, np.int8)
+            return [
+                spead2.send.HeapReference(gen_heap(timestamp, ant_index, frequency, data))
+                for ant_index in range(n_ants)
+                if present[ant_index]
+            ]
+
+        corrprod_results, beam_results, _, _ = await self._send_data(
+            mock_recv_streams,
+            mock_send_stream,
+            corrprod_outputs,
+            beam_outputs,
+            heap_factory=heap_factory,
+            present=np.ones((n_batches, n_ants), bool),
+            timestamp_step=timestamp_step,
+            n_channels_per_substream=n_channels_per_substream,
+            frequency=frequency,
+            n_spectra_per_heap=n_spectra_per_heap,
+        )
+
+        # NOTE: The results arrays returned are initialised as zeros, and should
+        # therefore still be zero for indices after the `?capture-stop` request
+        # was issued. That is, completely zero data is akin to no data received.
+        for stopped_stream in stopped_stream_names:
+            if stopped_stream in corrprod_results.keys():
+                # NOTE: As per the `DEFAULT_PARAMETERS`, both corrprod_outputs have the same
+                # heap_accumulation_threshold, which is why we can use either one to calculate
+                # the accumulation index where data stopped transmitting.
+                np.testing.assert_equal(
+                    corrprod_results[stopped_stream][
+                        capture_stop_chunk_id * HEAPS_PER_FENGINE_PER_CHUNK // heap_accumulation_threshold[0] :
```

Comment:
This line wrapping makes this whole statement hard to read. I'd suggest pulling the calculation into a separate statement (assigning it to a variable).